text/0000-user-ids.md

```diff
@@ -0,0 +1,44 @@
+# Secure runtime environments
+
+## Summary
+
+This RFC recommends that Paketo Buildpacks should in general strive to avoid making run-time modifications to the Buildpacks layers or the application directory by default. Instead, they should consider the output image to be read-only by default and explicitly set the parts that need to be writeable during runtime.
+
+
+## Motivation
+
+As of Buildpacks [RFC 0085](https://github.com/buildpacks/rfcs/blob/main/text/0085-run-uid.md) the Cloud Native Buildpacks project recommends that the run-time and build-time user ids be different. The motivation behind this RFC was to improve the security of output images and prevent arbitrary runtime modifications to built artifacts.
+
+The run-time and build-time user can potentially share the same group id if they wish to make modifications to the output image during run-time, by changing the permissions of the files they wish to modify during run-time to be group writable during the build process. Ideally most buildpacks should use the temporary directories or other known scratch directories for ephemeral runtime files.
+
+## Detailed Explanation
+
+This RFC recommends that we identify buildpacks that making runtime modifications to the application directory or buildpack layers and modify them to instead use a temporary directory - or have them make these files group writable instead. All the current Paketo stacks' build-time and run-time users belong to a unique group so changing the file permissions to be group writable should have no additional security implications for users using the Paketo stacks.
+
+
+## Rationale and Alternatives
+
+
+The rationale behind this RFC is explained in the upstream RFC - [RFC 0085](https://github.com/buildpacks/rfcs/blob/main/text/0085-run-uid.md)
+
+In the future we could also modify the Paketo stacks to have different run-time/build-time user ids once all the buildpacks follow this practice for a more secure experience.
+
+## Implementation
+
+At a high level this would involve the following - 
+
+- Create base stacks equivalent to the current Paketo stacks but with different runtime and build time user ids.
+- Create empty builders with the above stacks.
+- Run a parallel test pipeline which is optional and allowed to fail for each buildpack. This pipeline should utilize these new builders to gather the level of compliance for the current set of buildpacks. These tests should also provide appropriate directories for writing ephemeral files during runtime such as `/tmp` or `/run`. We can also use docker to mimic read only directories per https://www.thorsten-hans.com/read-only-filesystems-in-docker-and-kubernetes/
```

Comment:
What is the projected lifetime of this parallel test pipeline? It seems like a reasonable amount of overhead to have parallel pipelines. I'm not sure how others feel about that.

If it's not something we all need to do in the same way, I would prefer to just use this stricter running environment by default with the Java buildpacks. If anything breaks, then we'll fix it. That way we're not having a second pipeline specific to this task and we can continue to validate this is working in the future as well.

We're at around 30m running time for the present set of smoke tests for Java apps and we're talking about adding more tests. I don't really want to duplicate that and double the runtime, if possible.

Reply:
We could update existing tests to run with a different UID even if that means changing the user at runtime to be different than the default on our existing run images. We can make the change for real in the next major version of the stacks (`focal`??)